postgres/datadog_checks/postgres/statements.py

```diff
@@ -162,7 +172,12 @@ def row_keyfunc(row):
             return (row['query_signature'], row['datname'], row['rolname'])
 
         rows = self._normalize_queries(rows)
-        rows = self._state.compute_derivative_rows(rows, PG_STAT_STATEMENTS_METRICS_COLUMNS, key=row_keyfunc)
+        if len(rows) == 0:
+            return None
+
+        available_columns = set(rows[0].keys()).intersection(self._stat_column_cache)
+        metric_columns = available_columns.difference(PG_NON_MONOTONIC_COLUMNS)
```

Comment:
Actually this does look a bit dangerous to me. Because compute_derivative_rows performs subtraction, a newer postgres version (even minor version) could add a new text-based column that gets included in this list (and throw an error). I think I would prefer an allowlist for the metric columns after all @djova.

side note: you can use `+` and `-` operators on sets instead of the `.difference` methods, which tend to be more readable and intuitive.

Reply:
@justiniso This may not be as scary as you think, but I may not be seeing what you see. To explain, the `self._stat_column_cache` you see contains all the stat statement columns that we want, nothing more, which have the appropriate types (so it won't include strings, etc) and we just take the intersection of that to exclude all the other misc fields with types we don't want.

Reply:
Yeah if we want to make it truly dynamic then we need to filter out all non-number columns before doing the arithmetic.

An allowlist that contains all columns from all versions we support would be fine as well.

Reply:
The reason I say it's scary is because `available_columns` is doing `select * from {view}`. While that's not going to change at runtime, a customer could point to a newer postgres version we haven't explicitly supported yet. We 're aware of what the columns are in the latest version, but the next minor or major version of postgres/pg_stat_statments may release new columns like:

| column | type | example|
|---------|-----|--------|
| last_seen  |   int | 1621362328 |
| stddev_rows | int | 4004 |
| schema | string | botr |
| sampled_date | string | 2021-05-17 00:45:32.1234 |

These are unknown fields, so should they be included in the call to `compute_derivative_rows`? In this implementation, the integration would throw an error because the bottom two columns would mean subtracting strings from strings.

We could apply a filter for numeric values only; but then the resulting values would be gibberish for last_seen and stddev_rows because the numeric values are not monotonic counters. Maybe that's fine because we wouldn't use them right away, but then we would need to send the agent version back to know which payloads have correct values and which don't.

To me, a blocklist (versus allowlist) feels like a ticking time bomb that will come back to bite us as long as we are diffing monotonic counters.

Reply:
Valid point, I'm in favor of the allow list as well, I'll implement it now.